### PR TITLE
[REJECTED] REPL: Adjust default level of linting

### LINKED ISSUE #11888: import annotation._ is a drag for Seth
Subj.

Don't repackage, just set `-Yimports` to include it.

Also, `import util.chaining._` is a drag.

### PR DESCRIPTION
**Pull request by som-snytt**
Anything goes in REPL.

Fixes scala/bug#11888

### CODE REVIEW

File: src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala [hunk 1]
```diff
@@ -78,10 +78,51 @@ class MainGenericRunner {
...
-          // If user is agnostic about both -feature and -deprecation, turn them on.
+          // If user is agnostic about both -feature and -deprecation, turn them on and add batteries.
           if (settings.deprecation.isDefault && settings.feature.isDefault && settings.lint.isDefault) {
             settings.deprecation.value = true
             settings.feature.value = true
+            locally {
+              import settings.LintWarnings._
+              for (linted <- allLintWarnings if linted != Unused)
+                settings.lint.enable(linted)
+            }
```
Comment (som-snytt) [inline src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:89]: @lrytz sometimes I think the multichoice setting is clunky, but look ma, no strings!

File: src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala [hunk 2]
```diff
@@ -78,10 +78,51 @@ class MainGenericRunner {
...
+            locally {
+              import settings.UnusedWarnings._
+              val unused = List (
+                //Imports,  // not until REPL is fixed not to include spurious imports
```
Comment (dwijnand) [inline src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:93]: Even still.  An import that goes unused for a second shouldn't warn in a REPL, IMO.

And to be honest I'm not sure about adding and warning about all these things.  REPL code isn't production code.  If I define a pat var I don't end up immediately or even ever using, I don't want to be "yelled at" by the REPL about it.

IMO let's add the things we feel are just misleading not to warn about.  Or just add the imports, not the lints.

Comment (som-snytt) [reply src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:93]: In a REPL session, I am evolving some code. How to choose between "don't warn me, I'm just experimenting" and "please warn me so I evolve in the right direction"?

I'm leaning toward warning, so long as the presentation isn't verbose or interferes with normal session.

Comment (dwijnand) [reply src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:93]: Hmm, I see what you mean.  But maybe even when you're evolving some code, the REPL pointing out a bunch of lints is maybe too much.  Maybe removing the lints is a refinement process that you do after when you've move the code?

I'm not against adding enabling any lint warnings, but I don't think we have much room on change how verbose the presentation is.  I mean, unless we do something like make

```
scala> import scala.concurrent.Future
                               ^
       warning: Unused import
import scala.concurrent.Future
```

something like

```
scala> import scala.concurrent.Future
import scala.concurrent.Future (***)
```

Or do a warn once per session (with instructions for more info).

Comment (lrytz) [reply src/repl-frontend/scala/tools/nsc/MainGenericRunner.scala:93]: > How to choose between "don't warn me, I'm just experimenting" and "please warn me so I evolve in the right direction"?

It's a tough call, but I think I'm with Dale. Let's see which lint / unused / goodie warnings are clear wins and have low risk for being noisy even when writing drafty code, which people certainly do a lot in the REPL.

**Comment by som-snytt** [comment]
The bigger win will be warn once only on feature warnings, which I will add here.

I'll exercise this configuration and report. However, my hypothesis is that linting is not just about clean production code. It also helps a newbie understand why the code they're trying out doesn't work.

**Comment by som-snytt** [comment]
Sample SO question where linting helps if only people used it:
```
scala> val i: Integer = null ; Option(i): Option[Int]
                                     ^
       warning: Suspicious application of an implicit view (scala.Predef.Integer2int) in the argument to Option.apply.
i: Integer = null
res1: Option[Int] = Some(0)
```

- `:power up` facility
- warn once per session for language features
- report the true nature of a snippet
- `?` command to replay with all the lights on

**Comment by som-snytt** [comment]
Shout out to retronym for the lint at https://github.com/scala/bug/issues/6567, and to lrytz for nowarn so they can lint all the things.

**Comment by SethTisue** [comment]
(@som-snytt is this ready for me to look at?)

**Comment by som-snytt** [comment]
@SethTisue even I haven't looked at it yet.

I was going to get back to it soon. Somewhere I mentioned the problem of handling transient messages during edit cycle. The other dimension is that the whole industry of changing options in sbt for console, which I don't grasp, makes me want either a fix like this or a tpolecat plugin that will progress your options correctly.

My recent joke was do Scala devs in the UK know more about the succession of the monarchs or the succession of Scala options?

**Comment by som-snytt** [comment]
This is a useful future option for when repl is more useful.

**Pull request closed without merging**